<!DOCTYPE html>


  <html class="light page-post">


<head>
  <meta charset="utf-8">
  
  <title>rMATS原理及使用 | Hugo&#39;s Note</title>

  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">

  
    <meta name="keywords" content="RNA-Seq," />
  

  <meta name="description" content="rMATS，文章名Robust and flexible detection of differential alternative splicing from replicate RNA-Seq data，冲着这个replicate，我们就知道这是一个针对具有生物学重复开发的软件了，思路和RNA-Seq的差异基因检测类似，通过重复样本使用二项分布模型估算离散程度（或者不确定性），从而更好地校正数">
<meta property="og:type" content="article">
<meta property="og:title" content="rMATS原理及使用">
<meta property="og:url" content="https://fenghuji.github.io/2020/01/11/2020-1-11-rMATS%E5%8E%9F%E7%90%86%E5%8F%8A%E4%BD%BF%E7%94%A8/index.html">
<meta property="og:site_name" content="Hugo&#39;s Note">
<meta property="og:description" content="rMATS，文章名Robust and flexible detection of differential alternative splicing from replicate RNA-Seq data，冲着这个replicate，我们就知道这是一个针对具有生物学重复开发的软件了，思路和RNA-Seq的差异基因检测类似，通过重复样本使用二项分布模型估算离散程度（或者不确定性），从而更好地校正数">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/1.png">
<meta property="og:image" content="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/2.png">
<meta property="og:image" content="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/3.png">
<meta property="og:image" content="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/6.png">
<meta property="og:image" content="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/4.png">
<meta property="og:image" content="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/5.png">
<meta property="article:published_time" content="2020-01-11T05:33:40.000Z">
<meta property="article:modified_time" content="2020-01-11T05:33:40.000Z">
<meta property="article:author" content="Hugo Ji">
<meta property="article:tag" content="RNA-Seq">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/1.png">

  

  
    <link rel="icon" href="/favicon.ico">
  

  <link href="/css/styles.css?v=c114cbeddx" rel="stylesheet">


  
    
<link rel="stylesheet" href="/css/personal-style.css">

  

  
<!-- Google Analytics -->
<script type="text/javascript">
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
})(window,document,'script','//www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-154715039-1', 'auto');
ga('send', 'pageview');

</script>
<!-- End Google Analytics -->


  

  


  
    <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <link rel="stylesheet" href="//cdn.bootcss.com/font-awesome/4.3.0/css/font-awesome.min.css">
  

<!-- hexo injector head_end start -->
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.12.0/dist/katex.min.css">

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/hexo-math@4.0.0/dist/style.css">
<!-- hexo injector head_end end --><meta name="generator" content="Hexo 5.3.0"></head>

<body>


  
    <span id="toolbox-mobile" class="toolbox-mobile">盒子</span>
  

  <div class="post-header CENTER">
   
  <div class="toolbox">
    <a class="toolbox-entry" href="/">
      <span class="toolbox-entry-text">盒子</span>
      <i class="icon-angle-down"></i>
      <i class="icon-home"></i>
    </a>
    <ul class="list-toolbox">
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/archives/"
            rel="noopener noreferrer"
            target="_self"
            >
            博客
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/category/"
            rel="noopener noreferrer"
            target="_self"
            >
            分类
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/tag/"
            rel="noopener noreferrer"
            target="_self"
            >
            标签
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/link/"
            rel="noopener noreferrer"
            target="_self"
            >
            荐读
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/about/"
            rel="noopener noreferrer"
            target="_self"
            >
            关于
          </a>
        </li>
      
    </ul>
  </div>


</div>


  <div id="toc" class="toc-article">
    <strong class="toc-title">Posts List</strong>
    <ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E7%AE%80%E5%8D%95%E4%BB%8B%E7%BB%8D"><span class="toc-text">简单介绍</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Inclusion-Level%E7%9A%84%E8%AE%A1%E7%AE%97-Paired-Replicates"><span class="toc-text">Inclusion Level的计算(Paired Replicates)</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%85%B6%E4%BB%96%E7%B1%BB%E5%9E%8B%E7%9A%84%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87%E8%AE%A1%E7%AE%97"><span class="toc-text">其他类型的可变剪切计算</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%BD%AF%E4%BB%B6%E4%BD%BF%E7%94%A8"><span class="toc-text">软件使用</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%B8%8B%E8%BD%BD%E5%92%8C%E4%BE%9D%E8%B5%96"><span class="toc-text">下载和依赖</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%BD%BF%E7%94%A8%E5%8F%82%E6%95%B0"><span class="toc-text">使用参数</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E7%BB%93%E6%9E%9C%E6%96%87%E4%BB%B6"><span class="toc-text">结果文件</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Reference"><span class="toc-text">Reference</span></a></li></ol>
  </div>



<div class="content content-post CENTER">
   <article id="post-2020-1-11-rMATS原理及使用" class="article article-type-post" itemprop="blogPost">
  <header class="article-header">
    <h1 class="post-title">rMATS原理及使用</h1>

    <div class="article-meta">
      <span>
        <i class="icon-calendar"></i>
        <span>2020.01.11</span>
      </span>

      
        <span class="article-author">
          <i class="icon-user"></i>
          <span>Hugo Ji</span>
        </span>
      

      
  <span class="article-category">
    <i class="icon-list"></i>
    <a class="article-category-link" href="/categories/Bioinformatics/">Bioinformatics</a>
  </span>



      

      
      <i class="fa fa-eye"></i> 
        <span id="busuanzi_container_page_pv">
           &nbsp热度 <span id="busuanzi_value_page_pv">
           <i class="fa fa-spinner fa-spin"></i></span>℃
        </span>
      
      
    </div>
  </header>

  <div class="article-content">
    
      <p>rMATS，文章名<em>Robust and flexible detection of differential alternative splicing from replicate RNA-Seq data</em>，冲着这个<strong>replicate</strong>，我们就知道这是一个针对具有生物学重复开发的软件了，思路和RNA-Seq的差异基因检测类似，通过重复样本使用二项分布模型估算离散程度（或者不确定性），从而更好地校正数据，获取差异可变剪切事件，14-19年共400+引用，算是比较流行的软件。<br>由于分析需要，在此对rMATS的原理和使用简单记录。</p>
<h2 id="简单介绍"><a href="#简单介绍" class="headerlink" title="简单介绍"></a>简单介绍</h2><p>rMATS是一款对RNA-Seq数据进行差异可变剪切分析的软件，论文刊发的创新点在于针对具有生物重复的数据，构建了相匹配的模型来估计样本的离散特性，从而更好更可信的估计差异可变剪切事件，最后使用likelihood-ratio test来计算两组样品在IncLevel（Inclusion Level）水平上的差异，计算结果和PSI的定义相类型。然后计算校正的FDR值，将各种可变剪切结果的IncLevel值及差异FDR返回。</p>
<p>rMATS计算的可变剪切类型为主流5种：</p>
<ul>
<li>skipped exon (SE)外显子跳跃</li>
<li>alternative 5’ splice site (A5SS)第一个外显子可变剪切</li>
<li>alternative 3’ splice site (A3SS)最后一个外显子可变剪切</li>
<li>mutually exclusive exons (MXE)外显子选择性跳跃</li>
<li>retained intron (RI)内含子滞留</li>
</ul>
<p>图示为：</p>
<div style="text-align: center">
<img src="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/1.png"/>
</div>
rMATS使用一个likelihood-ratio test来计算P value来估计是否在两组数据中某个可变剪切事件的差异超过5%，即:
$$
|\Delta \psi|=\left|\psi_{i 1}-\psi_{i 2}\right|>5 \%
$$

其中$ \psi $代表某个isoform的外显子inclusion level，含义类似于常规可变剪切的PSI。
这种检测相对于直接计算两组的数据均值是否相等（差值为0）具有更好的统计显著性，需要注意的是，rMATS考虑了可变剪切事件的长度并进行标准化，而且计算的$ \psi $很大程度上和比例类似。

## Inclusion Level的计算(Unpaired Replicates)
我们以Skipped Exon的计算为例，针对下图
<div style="text-align: center">
<img src="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/2.png"/>
</div>
我们有两种事件分别为：exon inclusion isoform ($ I $)和exon skipping isoform($ S $)，中间的外显子被连接进去就是$ I $，长度为$ l_i $,跳过则为$ S $,长度为$ l_s $，对应的inclusion level $ \psi $计算公式为：
$$
\hat{\psi}=\left(I / l_{I}\right) /\left(I / l_{I}+S / l_{S}\right)
$$

**rMATS**假设$I$对应的count服从二项分布，总read count $n = I+S$,有：
$$
I | \psi \sim \text { Binomial }\left(n=I+S, p=f(\psi)=\frac{l_{l} \psi}{l_{l} \psi+l_{S}(1-\psi)}\right)
$$
那么$ \psi $就受到read count的影响，inclusion isoforms是经过有效长度标准化过的。
在生物重复实验中，组1有$ M_1 $样本，组2有$ M_2 $样本，推导有：
$$
\operatorname{logit}\left(\psi_{i j k}\right) \sim \operatorname{Normal}\left(\mu=\operatorname{logit}\left(\psi_{i j}\right), \sigma^{2}=\sigma_{i j}^{2}\right)
$$

上文即解释了rMATS对于生物重复的预估离散度或者说不确定性的估计（estimation uncertainty），使用**likelihood-ratio test**来计算两者的均值差是否超过阈值。一张图表示就是：
<div style="text-align: center">
<img src="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/3.png"/>
</div>

<h2 id="Inclusion-Level的计算-Paired-Replicates"><a href="#Inclusion-Level的计算-Paired-Replicates" class="headerlink" title="Inclusion Level的计算(Paired Replicates)"></a>Inclusion Level的计算(Paired Replicates)</h2><p>对于配对数据，在估计不确定的时候采取的是二元正态分布模型，具体计算细节比较复杂，在文章的附件可以仔细阅读，大体针对每个生物重复：</p>
<div style="text-align: center">
<img src="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/6.png"/>
</div>
式子中用到了方差和协相关系数，通过这样的假设模型来估计不确定性，一张图表示为：
<div style="text-align: center">
<img src="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/4.png"/>
</div>

<h2 id="其他类型的可变剪切计算"><a href="#其他类型的可变剪切计算" class="headerlink" title="其他类型的可变剪切计算"></a>其他类型的可变剪切计算</h2><p>文中介绍的是对于<strong>外显子跳跃</strong>，以此为例子，那么其他的可变剪切计算方法可以在文章附件中找到，见图：</p>
<div style="text-align: center">
<img src="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/rmats/5.png"/>
</div>

<h2 id="软件使用"><a href="#软件使用" class="headerlink" title="软件使用"></a>软件使用</h2><p>具体使用参考<a href="[https://www.bioinfo-scrounger.com/archives/346/](https://www.bioinfo-scrounger.com/archives/346/">rMATS差异可变剪切分析</a>)这篇博文或者<a target="_blank" rel="noopener" href="http://rnaseq-mats.sourceforge.net/user_guide.htm">rMATS Document</a>。<br><a target="_blank" rel="noopener" href="https://sourceforge.net/projects/rnaseq-mats/files/MATS/rMATS.4.0.1.tgz/download">下载网站</a>在这里，需要安装一些python库<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># For CentOS 6:</span></span><br><span class="line">pip install numpy</span><br><span class="line">yum install lapack-devel blas-devel</span><br><span class="line">yum install gsl-devel.x86_64</span><br><span class="line">yum install gcc-gfortran</span><br><span class="line"></span><br><span class="line"><span class="comment"># For Ubuntu 14:</span></span><br><span class="line">pip install numpy</span><br><span class="line">sudo apt-get install libblas-dev liblapack-dev</span><br><span class="line">sudo apt-get install libgsl0ldbl</span><br><span class="line">sudo apt-get install gfortran</span><br><span class="line"></span><br><span class="line"><span class="comment"># For Mac OS X Yosemite 10.10.5 (Using Homebrew for package management):</span></span><br><span class="line">brew install gcc@5</span><br><span class="line">brew install gsl</span><br><span class="line">pip install numpy</span><br></pre></td></tr></table></figure><br>新版rMATS下载解压后，你会发现有两个rmats.py执行脚本，这是由于rMATS v4.0.1 (turbo) was built with two different settings of Python interpreter，所以我们需要先测试下自己的系统支持那种，进入python，输入下述命令<br><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="built_in">print</span> sys.maxunicode</span><br></pre></td></tr></table></figure></p>
<ul>
<li>出现1114111则说明需要使用rMATS-turbo-Linux-UCS4文件夹下rmats.py</li>
<li>出现65535则说明使用rMATS-turbo-Linux-UCS2文件夹下rmats.py</li>
</ul>
<h3 id="下载和依赖"><a href="#下载和依赖" class="headerlink" title="下载和依赖"></a>下载和依赖</h3><ul>
<li>下载<a target="_blank" rel="noopener" href="http://rnaseq-mats.sourceforge.net/rMATS.4.0.2.tgz">rMATS v4.0.2 (turbo)</a></li>
<li>获得STAR genome indexes或者自己构建STAR索引</li>
<li>解压缩，进行分析</li>
</ul>
<h3 id="使用参数"><a href="#使用参数" class="headerlink" title="使用参数"></a>使用参数</h3><blockquote>
<p>–b1 b1.txt 输入sample1的txt格式的文件，文件内以逗号分隔重复样本的bam文件名<br>–b2 b2.txt 输入sample2的txt格式的文件，文件内以逗号分隔重复样本的bam文件名<br>-t readType 双端测序则readType为paired，单端测序则为single<br>–readLength 测序reads的长度<br>–gtf gtfFile 需要输入的gtf文件<br>–od outDir 所有输出文件的路径（文件夹）<br>–nthread 设置线程数<br>–cstat 上文的阈值c，代表的是两个样本中inclusion level的差值，0.1表示两个样本中该可变剪切事件的inclusion level相差10%</p>
</blockquote>
<p>bam文件需要通过比对得到，对于RNA数据而言，STAR的2-pass Mode是比较好的。</p>
<p>示例命令<br><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">python rMATS-turbo-Linux-UCS4&#x2F;rmats.py --b1 b1.txt \</span><br><span class="line">--b2 b2.txt --gtf Homo_sapiens.Ensembl.GRCh37.72.gtf \</span><br><span class="line">--od AS -t paired --readLength 151 --cstat 0.0001 --nthread 10</span><br></pre></td></tr></table></figure></p>
<h3 id="结果文件"><a href="#结果文件" class="headerlink" title="结果文件"></a>结果文件</h3><p>rMATS的结果文件是以各个可变剪切事件的分布的，主要由AS_Event.MATS.JC.txt，AS_Event.MATS.JCEC.txt，fromGTF.AS_Event.txt，JC.raw.input.AS_Event.txt，JCEC.raw.input.AS_Event.txt这几类；其中JC和JCEC的区别在于前者计算中使用的长度是junction length，而后者考虑junction length 和exon length，但是我们一般使用JC的结果就够了（如果只是单纯的比较两组样品间可变剪切的差异的话）。具体的长度计算见上文的<code>其他类型的可变剪切计算</code>。</p>
<p>下面几列信息比较重要：</p>
<ul>
<li>lncFormLen ：length of inclusion form, used for normalization</li>
<li>SkipFormLen : length of skipping form, used for normalization</li>
<li>P-Value : Significance of splicing difference between two sample groups（两组样品可变剪切的统计学显著差异指标）</li>
<li>FDR ： False Discovery Rate calculated from p-value（对p-value的FDR校正）</li>
<li>lncLevel1 ： inclusion level for SAMPLE_1 replicates (comma separated) calculated from normalized counts</li>
<li>IncLevel2 : inclusion level for SAMPLE_2 replicates (comma separated) calculated from normalized counts</li>
<li>IncLevelDifference ： average(IncLevel1) - average(IncLevel2)</li>
</ul>
<p>那实际上很多时候我们也就看看FDR，还有一些坐标信息有时候也是我们需要的。</p>
<h2 id="Reference"><a href="#Reference" class="headerlink" title="Reference"></a>Reference</h2><p><a href="rMATS差异可变剪切分析">rMATS差异可变剪切分析</a><br><a target="_blank" rel="noopener" href="https://baike.baidu.com/item/%E9%80%89%E6%8B%A9%E6%80%A7%E5%89%AA%E6%8E%A5/3648412?fromtitle=%E5%8F%AF%E5%8F%98%E5%89%AA%E6%8E%A5&amp;fromid=277930">可变剪切百度百科</a><br><a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pubmed/25480548">文章原文及附件</a><br><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/98bb7a7d6077">使用rmats进行可变剪切的分析</a><br><a target="_blank" rel="noopener" href="http://rnaseq-mats.sourceforge.net/user_guide.htm">rMATS Document</a></p>

    
  </div>

</article>


   
  <div class="text-center donation">
    <div class="inner-donation">
      <span class="btn-donation">支持一下</span>
      <div class="donation-body">
        <div class="tip text-center">自愿投币</div>
        <ul>
        
          <li class="item">
            
              <span>支付宝扫一扫</span>
            
            <img src="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/%E4%BA%8C%E7%BB%B4%E7%A0%81.jpg" alt="">
          </li>
        
        </ul>
      </div>
    </div>
  </div>


   
  <div class="box-prev-next clearfix">
    <a class="show pull-left" href="/2020/01/08/%E8%B5%B0%E5%90%912020%E5%B9%B4%EF%BC%8C%E7%BB%99%E8%87%AA%E5%B7%B1%E4%B8%80%E4%BA%9B%E5%AF%84%E8%AF%AD/">
        <i class="icon icon-angle-left"></i>
    </a>
    <a class="show pull-right" href="/2020/01/13/(%E7%AC%94%E8%AE%B0%E5%9B%9B)ggplot2%E5%B7%A5%E5%85%B7%E7%AE%B1/">
        <i class="icon icon-angle-right"></i>
    </a>
  </div>




</div>


  <a id="backTop" class="back-top">
    <i class="icon-angle-up"></i>
  </a>




  <div class="modal" id="modal">
  <span id="cover" class="cover hide"></span>
  <div id="modal-dialog" class="modal-dialog hide-dialog">
    <div class="modal-header">
      <span id="close" class="btn-close">Close</span>
    </div>
    <hr>
    <div class="modal-body">
      <ul class="list-toolbox">
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/archives/"
              rel="noopener noreferrer"
              target="_self"
              >
              博客
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/category/"
              rel="noopener noreferrer"
              target="_self"
              >
              分类
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/tag/"
              rel="noopener noreferrer"
              target="_self"
              >
              标签
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/link/"
              rel="noopener noreferrer"
              target="_self"
              >
              荐读
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/about/"
              rel="noopener noreferrer"
              target="_self"
              >
              关于
            </a>
          </li>
        
      </ul>

    </div>
  </div>
</div>



  
      <div class="fexo-comments comments-post">
    

    

    
    

    

    
    

    

<!-- Gitalk评论插件通用代码 -->
<div id="gitalk-container"></div>

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.css">
<script src="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js"></script>
<script src="//cdn.bootcss.com/blueimp-md5/2.10.0/js/md5.js"></script>
<script>
const gitalk = new Gitalk({
  clientID: '0963ec29d0c3a87ab983',
  clientSecret: 'd9c867a5cb0ee1dc63198451aa8ba3c80d865af9',
  repo: 'fenghuji.github.io',
  owner: 'fenghuji',
  // 在这里设置一下截取前50个字符串, 这是因为 github 对 label 的长度有了要求, 如果超过
  // 50个字符串则会报错.
  // id: location.pathname.split('/').pop().substring(0, 49),
  id: md5(location.pathname),
  admin: ['fenghuji'],
  // facebook-like distraction free mode
  distractionFreeMode: false
})
gitalk.render('gitalk-container')
</script>
<!-- Gitalk代码结束 -->



  </div>

  

  <script type="text/javascript">
  function loadScript(url, callback) {
    var script = document.createElement('script')
    script.type = 'text/javascript';

    if (script.readyState) { //IE
      script.onreadystatechange = function() {
        if (script.readyState == 'loaded' ||
          script.readyState == 'complete') {
          script.onreadystatechange = null;
          callback();
        }
      };
    } else { //Others
      script.onload = function() {
        callback();
      };
    }

    script.src = url;
    document.getElementsByTagName('head')[0].appendChild(script);
  }

  window.onload = function() {
    loadScript('/js/bundle.js?235683', function() {
      // load success
    });
  }
</script>

<script src="/live2dw/lib/L2Dwidget.min.js?094cbace49a39548bed64abff5988b05"></script><script>L2Dwidget.init({"pluginModelPath":"assets/","model":{"jsonPath":"/live2dw/assets/z16.model.json"},"display":{"position":"left","width":500,"height":1000},"mobile":{"show":false},"rect":"opacity:0.7","log":false,"pluginJsPath":"lib/","pluginRootPath":"live2dw/","tagMode":false});</script></body>
</html>

<canvas class="fireworks" style="position: fixed;left: 0;top: 0;z-index: 1; pointer-events: none;" ></canvas> 
<script type="text/javascript" src="//cdn.bootcss.com/animejs/2.2.0/anime.min.js"></script> 
<script type="text/javascript" src="/js/fireworks.js"></script>

