<!DOCTYPE html>


  <html class="light page-post">


<head>
  <meta charset="utf-8">
  
  <title>edgeR的使用 | Hugo&#39;s Note</title>

  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">

  
    <meta name="keywords" content="RNA-Seq,R,edgeR," />
  

  <meta name="description" content="edgeR常见问题edgeR是差异表达分析常见的方法，最近需要明白其原理，所以深入学习了一下，对其中一些需要注意的地方记录下来。  edgeR是基于read counts的整数表格工作的，其中行对应gene，列对应每一个独立的文库，counts对应每个gene比对上的reads数目。 因为edgeR是对实际的read count进行设计的，所以不推荐使用预测的转录本丰度作为输入来计算。 edgeR">
<meta property="og:type" content="article">
<meta property="og:title" content="edgeR的使用">
<meta property="og:url" content="https://fenghuji.github.io/2019/11/14/edgeR/index.html">
<meta property="og:site_name" content="Hugo&#39;s Note">
<meta property="og:description" content="edgeR常见问题edgeR是差异表达分析常见的方法，最近需要明白其原理，所以深入学习了一下，对其中一些需要注意的地方记录下来。  edgeR是基于read counts的整数表格工作的，其中行对应gene，列对应每一个独立的文库，counts对应每个gene比对上的reads数目。 因为edgeR是对实际的read count进行设计的，所以不推荐使用预测的转录本丰度作为输入来计算。 edgeR">
<meta property="og:locale" content="en_US">
<meta property="article:published_time" content="2019-11-13T16:00:00.000Z">
<meta property="article:modified_time" content="2019-11-13T16:00:00.000Z">
<meta property="article:author" content="Hugo Ji">
<meta property="article:tag" content="RNA-Seq">
<meta property="article:tag" content="R">
<meta property="article:tag" content="edgeR">
<meta name="twitter:card" content="summary">

  

  
    <link rel="icon" href="/favicon.ico">
  

  <link href="/css/styles.css?v=c114cbeddx" rel="stylesheet">


  
    
<link rel="stylesheet" href="/css/personal-style.css">

  

  
<!-- Google Analytics -->
<script type="text/javascript">
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
})(window,document,'script','//www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-154715039-1', 'auto');
ga('send', 'pageview');

</script>
<!-- End Google Analytics -->


  

  


  
    <script async src="https://busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
    <link rel="stylesheet" href="//cdn.bootcss.com/font-awesome/4.3.0/css/font-awesome.min.css">
  

<meta name="generator" content="Hexo 5.3.0"></head>

<body>


  
    <span id="toolbox-mobile" class="toolbox-mobile">盒子</span>
  

  <div class="post-header CENTER">
   
  <div class="toolbox">
    <a class="toolbox-entry" href="/">
      <span class="toolbox-entry-text">盒子</span>
      <i class="icon-angle-down"></i>
      <i class="icon-home"></i>
    </a>
    <ul class="list-toolbox">
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/archives/"
            rel="noopener noreferrer"
            target="_self"
            >
            博客
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/category/"
            rel="noopener noreferrer"
            target="_self"
            >
            分类
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/tag/"
            rel="noopener noreferrer"
            target="_self"
            >
            标签
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/link/"
            rel="noopener noreferrer"
            target="_self"
            >
            荐读
          </a>
        </li>
      
        <li class="item-toolbox">
          <a
            class="CIRCLE"
            href="/about/"
            rel="noopener noreferrer"
            target="_self"
            >
            关于
          </a>
        </li>
      
    </ul>
  </div>


</div>


  <div id="toc" class="toc-article">
    <strong class="toc-title">Posts List</strong>
    <ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#edgeR%E5%B8%B8%E8%A7%81%E9%97%AE%E9%A2%98"><span class="toc-text">edgeR常见问题</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%A0%87%E5%87%86%E5%8C%96%E9%97%AE%E9%A2%98"><span class="toc-text">标准化问题</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#edgeR%E5%AE%98%E6%96%B9%E6%95%99%E7%A8%8B"><span class="toc-text">edgeR官方教程</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%AF%BB%E5%85%A5%E6%95%B0%E6%8D%AE"><span class="toc-text">读入数据</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#Annotation"><span class="toc-text">Annotation</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%BF%87%E6%BB%A4%E5%92%8C%E6%A0%87%E5%87%86%E5%8C%96"><span class="toc-text">过滤和标准化</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%9F%A5%E7%9C%8B%E6%95%B0%E6%8D%AE"><span class="toc-text">查看数据</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90"><span class="toc-text">差异表达分析</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#GO%E6%B3%A8%E9%87%8A"><span class="toc-text">GO注释</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%88%A9%E7%94%A8egdeR%E5%BF%AB%E9%80%9F%E8%BF%9B%E8%A1%8C%E5%A4%9A%E7%A7%8D%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E6%96%B9%E6%B3%95%E7%9A%84%E4%BB%A3%E7%A0%81"><span class="toc-text">利用egdeR快速进行多种差异表达方法的代码</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#Reference"><span class="toc-text">Reference:</span></a></li></ol>
  </div>



<div class="content content-post CENTER">
   <article id="post-edgeR" class="article article-type-post" itemprop="blogPost">
  <header class="article-header">
    <h1 class="post-title">edgeR的使用</h1>

    <div class="article-meta">
      <span>
        <i class="icon-calendar"></i>
        <span>2019.11.14</span>
      </span>

      
        <span class="article-author">
          <i class="icon-user"></i>
          <span>Hugo Ji</span>
        </span>
      

      
  <span class="article-category">
    <i class="icon-list"></i>
    <a class="article-category-link" href="/categories/Bioinformatics/">Bioinformatics</a>
  </span>



      

      
      <i class="fa fa-eye"></i> 
        <span id="busuanzi_container_page_pv">
           &nbsp热度 <span id="busuanzi_value_page_pv">
           <i class="fa fa-spinner fa-spin"></i></span>℃
        </span>
      
      
    </div>
  </header>

  <div class="article-content">
    
      <h2 id="edgeR常见问题"><a href="#edgeR常见问题" class="headerlink" title="edgeR常见问题"></a>edgeR常见问题</h2><p>edgeR是差异表达分析常见的方法，最近需要明白其原理，所以深入学习了一下，对其中一些需要注意的地方记录下来。</p>
<ul>
<li>edgeR是<strong>基于read counts的整数表格</strong>工作的，其中行对应gene，列对应每一个独立的文库，counts对应每个gene比对上的reads数目。</li>
<li>因为edgeR是对实际的read count进行设计的，所以不推荐使用预测的转录本丰度作为输入来计算。</li>
<li>edgeR使用一个叫做DGEList的数据类(data class)来进行工作的，使用DGEList函数对一个counts矩阵或者数据框工作是非常快速的。</li>
<li>edgeR使用count-per-million (CPM)来进行数据过滤而不是reads数目本身，因为后者随文库的不同而具有差异。</li>
</ul>
<a id="more"></a>

<h3 id="标准化问题"><a href="#标准化问题" class="headerlink" title="标准化问题"></a>标准化问题</h3><ul>
<li>和实验技术无关的因素不需要被考虑进差异表达的问题中：比如read counts受基因长度影响，但是gene长度在每个样本中都有这样的影响，那就不需要考虑</li>
<li>edgeR对不同样本制备的测序深度进行了数据调整</li>
<li>GC含量和基因长度对分析没有太多影响</li>
<li>edgeR使用矫正因子代入模型来做标准化，矫正因子是edgeR函数内部计算的。其标准化是基于模型的，所以用户不要输入任何其他类型的丰度数据如RPKM或者FPKM，这会使得错误估计模型参数引入错误</li>
<li>edgeR可以输入<code>RSEM</code>产生的结果，但raw reads counts更好</li>
</ul>
<h2 id="edgeR官方教程"><a href="#edgeR官方教程" class="headerlink" title="edgeR官方教程"></a>edgeR官方教程</h2><p>RNA-Seq数据，口腔癌和正常样本的差异表达分析流程</p>
<h3 id="读入数据"><a href="#读入数据" class="headerlink" title="读入数据"></a>读入数据</h3><p>数据来自于文章<a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2824832/bin/pone.0009317.s009.xls">附表S1</a>整理而来，百度云网址为<a target="_blank" rel="noopener" href="https://pan.baidu.com/s/1Zvv0xc4O1KTRTvWOQp6Hjg">链接</a>, 提取码：0k2g </p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">rawdata &lt;- read.csv(<span class="string">&quot;TableS1.csv&quot;</span>)</span><br><span class="line">head(rawdata)</span><br><span class="line">library(edgeR)</span><br><span class="line">y &lt;- DGEList(counts=rawdata[,<span class="number">4</span>:<span class="number">9</span>], genes=rawdata[,<span class="number">1</span>:<span class="number">3</span>])</span><br></pre></td></tr></table></figure>
<p><code>head(rawdata)</code>的结果为:</p>
<table>
<thead>
<tr>
<th></th>
<th>idRefSeq</th>
<th>nameOfGene</th>
<th>numberOfExons</th>
<th>X8N</th>
<th>X8T</th>
<th>X33N</th>
<th>X33T</th>
<th>X51N</th>
<th>X51T</th>
</tr>
</thead>
<tbody><tr>
<td>1</td>
<td>NM_182502</td>
<td>TMPRSS11B</td>
<td>10</td>
<td>2592</td>
<td>3</td>
<td>7805</td>
<td>321</td>
<td>3372</td>
<td>9</td>
</tr>
<tr>
<td>2</td>
<td>NM_003280</td>
<td>TNNC1</td>
<td>6</td>
<td>1684</td>
<td>0</td>
<td>1787</td>
<td>7</td>
<td>4894</td>
<td>559</td>
</tr>
<tr>
<td>3</td>
<td>NM_152381</td>
<td>XIRP2</td>
<td>10</td>
<td>9915</td>
<td>15</td>
<td>10396</td>
<td>48</td>
<td>23309</td>
<td>7171</td>
</tr>
<tr>
<td>4</td>
<td>NM_022438</td>
<td>MAL</td>
<td>3</td>
<td>2496</td>
<td>2</td>
<td>3585</td>
<td>239</td>
<td>1596</td>
<td>7</td>
</tr>
<tr>
<td>5</td>
<td>NM_1100112</td>
<td>MYH2</td>
<td>40</td>
<td>4389</td>
<td>7</td>
<td>7944</td>
<td>16</td>
<td>9262</td>
<td>1818</td>
</tr>
<tr>
<td>6</td>
<td>NM_017534</td>
<td>MYH2</td>
<td>40</td>
<td>4402</td>
<td>7</td>
<td>7943</td>
<td>16</td>
<td>9244</td>
<td>1815</td>
</tr>
</tbody></table>
<h3 id="Annotation"><a href="#Annotation" class="headerlink" title="Annotation"></a>Annotation</h3><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">library(org.Hs.eg.db)</span><br><span class="line">idfound &lt;- y$genes$idRefSeq %in% mappedRkeys(org.Hs.egREFSEQ)</span><br><span class="line">y &lt;- y[idfound,]</span><br><span class="line"><span class="comment">#dim(y)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##加上Entrez Gene IDs注释</span></span><br><span class="line">egREFSEQ &lt;- toTable(org.Hs.egREFSEQ)</span><br><span class="line"><span class="comment">#head(egREFSEQ)</span></span><br><span class="line">m &lt;- match(y$genes$idRefSeq, egREFSEQ$accession)</span><br><span class="line">y$genes$EntrezGene &lt;- egREFSEQ$gene_id[m]</span><br><span class="line"></span><br><span class="line"><span class="comment">## 使用the Entrez Gene IDs更新Gene SYMBOL</span></span><br><span class="line">egSYMBOL &lt;- toTable(org.Hs.egSYMBOL)</span><br><span class="line"><span class="comment">#head(egSYMBOL)</span></span><br><span class="line">m &lt;- match(y$genes$EntrezGene, egSYMBOL$gene_id)</span><br><span class="line">y$genes$Symbol &lt;- egSYMBOL$symbol[m]</span><br><span class="line"><span class="comment">#head(y$genes)</span></span><br></pre></td></tr></table></figure>
<h3 id="过滤和标准化"><a href="#过滤和标准化" class="headerlink" title="过滤和标准化"></a>过滤和标准化</h3><p>不同的基因有多个转录本且表达量类似，我们选择一个gene的最高表达量转录本作为代表</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">o &lt;- order(rowSums(y$counts), decreasing=<span class="literal">TRUE</span>)</span><br><span class="line">y &lt;- y[o,]</span><br><span class="line">d &lt;- duplicated(y$genes$Symbol)</span><br><span class="line">y &lt;- y[!d,]</span><br><span class="line"><span class="comment">#nrow(y)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 过滤低表达量genes，在该转录本数据中已经设定了在每个样本中至少出现一次并具有至少50个reads</span></span><br><span class="line">y$samples$lib.size &lt;- colSums(y$counts)  <span class="comment">#重新计算文库大小</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#使用Entrez Gene IDs作为行名</span></span><br><span class="line">rownames(y$counts) &lt;- rownames(y$genes) &lt;- y$genes$EntrezGene</span><br><span class="line">y$genes$EntrezGene &lt;- <span class="literal">NULL</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#TMM normalization</span></span><br><span class="line">y &lt;- calcNormFactors(y)</span><br><span class="line"><span class="comment">#y$samples</span></span><br></pre></td></tr></table></figure>
<h3 id="查看数据"><a href="#查看数据" class="headerlink" title="查看数据"></a>查看数据</h3><p>检查数据的异常值，函数<code>plotMDS</code>可以查看样本间的距离并计算生物学变化系数<code>biological coefficient of variation (BCV)</code></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotMDS(y)  <span class="comment">##维度1将癌症和正常样本分离开来，维度2对应病人个数，癌症病人比正常人具有更高的异质性</span></span><br></pre></td></tr></table></figure>
<p>设置矩阵来及进行差异表达</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">Patient &lt;- factor(<span class="built_in">c</span>(<span class="number">8</span>,<span class="number">8</span>,<span class="number">33</span>,<span class="number">33</span>,<span class="number">51</span>,<span class="number">51</span>))</span><br><span class="line">Tissue &lt;- factor(<span class="built_in">c</span>(<span class="string">&quot;N&quot;</span>,<span class="string">&quot;T&quot;</span>,<span class="string">&quot;N&quot;</span>,<span class="string">&quot;T&quot;</span>,<span class="string">&quot;N&quot;</span>,<span class="string">&quot;T&quot;</span>))</span><br><span class="line">data.frame(Sample=colnames(y),Patient,Tissue)</span><br><span class="line">design &lt;- model.matrix(~Patient+Tissue)</span><br><span class="line">rownames(design) &lt;- colnames(y)</span><br><span class="line"><span class="comment">#design</span></span><br></pre></td></tr></table></figure>
<p>预估分散度：<br>负二项分布（negative binomial，NB）模型需要均值和离散值两个参数。首先需要根据样本分组，或者说根据实验设计、目的，构建一个分组矩阵。然后使用该分组矩阵，通过加权似然经验贝叶斯（weighted likelihood empirical Bayes）模型，估算每个基因的负二项离散Tagwise（即经验贝叶斯稳健离散值）、Common（即经验贝叶斯稳健离散值的均值）、Trended（即经验贝叶斯稳健离散值的拟合值）。<br>在edgeR中还有很多函数可以单独估算离散值，如<code>estimateCommonDisp()</code>、<code>estimateTagwiseDisp()</code>、<code>estimateGLMTagwiseDisp()</code>、<code>estimateGLMCommonDisp()</code>、<code>estimateGLMTrendedDisp()</code>等，具体可以参阅帮助文档。当然不难看出，我们使用的estimateDisp()实际上是个组合函数，可以一步得到多个计算结果，我们使用分组矩阵<code>design</code>通过<code>estimateDisp()</code>可以估算3个值，其实是<code>estimateGLMTagwiseDisp()</code>、<code>estimateGLMCommonDisp()</code>和<code>estimateGLMTrendedDisp()</code>这3个结果的组合。如果不指定分组矩阵，则将得到<code>estimateCommonDisp()</code>和<code>estimateTagwiseDisp()</code>的结果组合。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">y &lt;- estimateDisp(y, design, robust=<span class="literal">TRUE</span>)</span><br><span class="line">y$common.dispersion</span><br><span class="line">plotBCV(y)  <span class="comment"># common dispersion的平方根给出了biological variation的方差，这里common dispersion是0.159，那么biological variation系数约为0.4</span></span><br></pre></td></tr></table></figure>
<h3 id="差异表达分析"><a href="#差异表达分析" class="headerlink" title="差异表达分析"></a>差异表达分析</h3><p>genewise glms方法： 类似t检验，top DE tags 具有<strong>低p-value和FDR，高的fold changes</strong></p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">fit &lt;- glmFit(y, design)</span><br><span class="line">lrt &lt;- glmLRT(fit)</span><br><span class="line">topTags(lrt)</span><br><span class="line">summary(decideTests(lrt))</span><br><span class="line">plotMD(lrt)</span><br><span class="line">abline(h=<span class="built_in">c</span>(-<span class="number">1</span>, <span class="number">1</span>), col=<span class="string">&quot;blue&quot;</span>)  <span class="comment"># 蓝线代表2倍差异</span></span><br></pre></td></tr></table></figure>
<h3 id="GO注释"><a href="#GO注释" class="headerlink" title="GO注释"></a>GO注释</h3><p>针对*biological process (BP)*的生物注释，癌症上调genes和细胞分化，细胞转移和组织形态有关：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">go &lt;- goana(lrt)</span><br><span class="line">topGO(go, ont=<span class="string">&quot;BP&quot;</span>, sort=<span class="string">&quot;Up&quot;</span>, n=<span class="number">30</span>, truncate=<span class="number">30</span>)</span><br></pre></td></tr></table></figure>

<h2 id="利用egdeR快速进行多种差异表达方法的代码"><a href="#利用egdeR快速进行多种差异表达方法的代码" class="headerlink" title="利用egdeR快速进行多种差异表达方法的代码"></a>利用egdeR快速进行多种差异表达方法的代码</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## Differential expression analysis</span></span><br><span class="line"><span class="comment">## install R package:edgeR</span></span><br><span class="line"><span class="keyword">if</span> (!requireNamespace(<span class="string">&quot;BiocManager&quot;</span>, quietly = <span class="literal">TRUE</span>))</span><br><span class="line">    install.packages(<span class="string">&quot;BiocManager&quot;</span>)</span><br><span class="line">BiocManager::install(<span class="string">&quot;edgeR&quot;</span>)</span><br><span class="line">library(edgeR)</span><br><span class="line"></span><br><span class="line">setwd(<span class="string">&quot;G:\\Projects\\co-expression&quot;</span>)</span><br><span class="line">target&lt;-read.table(<span class="string">&quot;reads.txt&quot;</span>,header=<span class="built_in">T</span>,row.names=<span class="number">1</span>)</span><br><span class="line">group&lt;-<span class="built_in">c</span>(<span class="built_in">rep</span>(<span class="string">&quot;T&quot;</span>,<span class="number">24</span>),<span class="built_in">rep</span>(<span class="string">&quot;N&quot;</span>,<span class="number">15</span>))</span><br><span class="line">dgelist &lt;- DGEList(counts = target, group = group)</span><br><span class="line"></span><br><span class="line"><span class="comment">## filter gene with too low abundance</span></span><br><span class="line">keep &lt;- rowSums(cpm(dgelist) &gt; <span class="number">1</span> ) &gt;= <span class="number">2</span></span><br><span class="line">dgelist &lt;- dgelist[keep,keep.lib.sizes = <span class="literal">FALSE</span>]</span><br><span class="line"><span class="comment">## normalization</span></span><br><span class="line">dgelist_norm &lt;- calcNormFactors(dgelist, method = <span class="string">&#x27;TMM&#x27;</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#估算离散值</span></span><br><span class="line"><span class="comment">#install.packages(&quot;statmod&quot;)</span></span><br><span class="line">design &lt;- model.matrix(~group)    <span class="comment">#构建分组矩阵</span></span><br><span class="line">dge &lt;- estimateDisp(dgelist_norm, design, robust = <span class="literal">TRUE</span>) <span class="comment">#估算离散值</span></span><br><span class="line"><span class="comment">#plotBCV(dge) #作图查看</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#negative binomial generalized log-linear model</span></span><br><span class="line">fit &lt;- glmFit(dge, design, robust = <span class="literal">TRUE</span>)     <span class="comment">#拟合模型</span></span><br><span class="line">lrt &lt;- glmLRT(fit)   <span class="comment">#统计检验</span></span><br><span class="line">topTags(lrt)</span><br><span class="line">write.csv(topTags(lrt, n = nrow(dgelist$counts)), <span class="string">&#x27;glmLRT.csv&#x27;</span>, <span class="built_in">quote</span> = <span class="literal">FALSE</span>) <span class="comment">#输出主要结果</span></span><br><span class="line">dge_de &lt;- decideTestsDGE(lrt, adjust.method = <span class="string">&#x27;fdr&#x27;</span>, p.value = <span class="number">0.05</span>)  <span class="comment">#查看默认方法获得的差异基因</span></span><br><span class="line">summary(dge_de)</span><br><span class="line">plotMD(lrt, status = dge_de, values = <span class="built_in">c</span>(<span class="number">1</span>, -<span class="number">1</span>), col = <span class="built_in">c</span>(<span class="string">&#x27;blue&#x27;</span>, <span class="string">&#x27;red&#x27;</span>))     <span class="comment">#作图观测</span></span><br><span class="line">abline(h = <span class="built_in">c</span>(-<span class="number">1</span>, <span class="number">1</span>), col = <span class="string">&#x27;gray&#x27;</span>, lty = <span class="number">2</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#quasi-likelihood negative binomial generalized log-linear model 拟合</span></span><br><span class="line">fit &lt;- glmQLFit(dge, design, robust = <span class="literal">TRUE</span>)  <span class="comment">#拟合模型</span></span><br><span class="line">lrt &lt;- glmQLFTest(fit) <span class="comment">#统计检验</span></span><br><span class="line">topTags(lrt)</span><br><span class="line">write.csv(topTags(lrt, n = nrow(dgelist$counts)), <span class="string">&#x27;glmQLFTest.csv&#x27;</span>, <span class="built_in">quote</span> = <span class="literal">FALSE</span>)  <span class="comment">#输出主要结果</span></span><br><span class="line">dge_de &lt;- decideTestsDGE(lrt, adjust.method = <span class="string">&#x27;fdr&#x27;</span>, p.value = <span class="number">0.05</span>)  <span class="comment">#查看默认方法获得的差异基因</span></span><br><span class="line">summary(dge_de)</span><br><span class="line">plotMD(lrt, status = dge_de, values = <span class="built_in">c</span>(<span class="number">1</span>, -<span class="number">1</span>), col = <span class="built_in">c</span>(<span class="string">&#x27;blue&#x27;</span>, <span class="string">&#x27;red&#x27;</span>))   <span class="comment">#作图观测</span></span><br><span class="line">abline(h = <span class="built_in">c</span>(-<span class="number">1</span>, <span class="number">1</span>), col = <span class="string">&#x27;gray&#x27;</span>, lty = <span class="number">2</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#配对检验</span></span><br><span class="line">dge_et &lt;- exactTest(dge) <span class="comment">#检验</span></span><br><span class="line">topTags(dge_et)</span><br><span class="line">write.csv(topTags(dge_et, n = nrow(dgelist$counts)), <span class="string">&#x27;exactTest.csv&#x27;</span>, <span class="built_in">quote</span> = <span class="literal">FALSE</span>)  <span class="comment">#输出主要结果</span></span><br><span class="line">dge_de &lt;- decideTestsDGE(dge_et, adjust.method = <span class="string">&#x27;fdr&#x27;</span>, p.value = <span class="number">0.05</span>)   <span class="comment">#查看默认方法获得的差异基因</span></span><br><span class="line">summary(dge_de)</span><br><span class="line">detags &lt;- rownames(dge)[<span class="built_in">as.logical</span>(dge_de)]</span><br><span class="line">plotSmear(dge_et, de.tags = detags, cex = <span class="number">0.5</span>)      <span class="comment">#作图观测</span></span><br><span class="line">abline(h = <span class="built_in">c</span>(-<span class="number">1</span>, <span class="number">1</span>), col = <span class="string">&#x27;gray&#x27;</span>, lty = <span class="number">2</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#voom 线性建模（limma）</span></span><br><span class="line">limma_voom &lt;- voom(dgelist_norm, design, plot = <span class="literal">TRUE</span>)</span><br><span class="line">fit &lt;- lmFit(limma_voom, design)  <span class="comment">#拟合</span></span><br><span class="line">fit &lt;- eBayes(fit)</span><br><span class="line">topTable(fit, coef = <span class="number">2</span>)</span><br><span class="line">write.csv(topTable(fit, coef = <span class="number">2</span>, number = nrow(dgelist$counts)), <span class="string">&#x27;limma_voom.csv&#x27;</span>, <span class="built_in">quote</span> = <span class="literal">FALSE</span>)      <span class="comment">#输出主要结果</span></span><br></pre></td></tr></table></figure>
<h2 id="Reference"><a href="#Reference" class="headerlink" title="Reference:"></a>Reference:</h2><p><a target="_blank" rel="noopener" href="http://blog.sciencenet.cn/blog-3406804-1188480.html">R包edgeR、limma差异基因分析</a><br><a target="_blank" rel="noopener" href="https://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf">edgeRUsersGuide</a></p>

    
  </div>

</article>


   
  <div class="text-center donation">
    <div class="inner-donation">
      <span class="btn-donation">支持一下</span>
      <div class="donation-body">
        <div class="tip text-center">自愿投币</div>
        <ul>
        
          <li class="item">
            
              <span>支付宝扫一扫</span>
            
            <img src="https://hugo-blog-1256988836.cos.ap-chengdu.myqcloud.com/Blog/Images/%E4%BA%8C%E7%BB%B4%E7%A0%81.jpg" alt="">
          </li>
        
        </ul>
      </div>
    </div>
  </div>


   
  <div class="box-prev-next clearfix">
    <a class="show pull-left" href="/wgcna1">
        <i class="icon icon-angle-left"></i>
    </a>
    <a class="show pull-right" href="/2019/11/23/%E3%80%8A%E6%88%90%E4%BA%8B%E5%9C%A8%E5%A4%A9%E3%80%8B%E8%AF%BB%E5%90%8E%E6%84%9F/">
        <i class="icon icon-angle-right"></i>
    </a>
  </div>




</div>


  <a id="backTop" class="back-top">
    <i class="icon-angle-up"></i>
  </a>




  <div class="modal" id="modal">
  <span id="cover" class="cover hide"></span>
  <div id="modal-dialog" class="modal-dialog hide-dialog">
    <div class="modal-header">
      <span id="close" class="btn-close">Close</span>
    </div>
    <hr>
    <div class="modal-body">
      <ul class="list-toolbox">
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/archives/"
              rel="noopener noreferrer"
              target="_self"
              >
              博客
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/category/"
              rel="noopener noreferrer"
              target="_self"
              >
              分类
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/tag/"
              rel="noopener noreferrer"
              target="_self"
              >
              标签
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/link/"
              rel="noopener noreferrer"
              target="_self"
              >
              荐读
            </a>
          </li>
        
          <li class="item-toolbox">
            <a
              class="CIRCLE"
              href="/about/"
              rel="noopener noreferrer"
              target="_self"
              >
              关于
            </a>
          </li>
        
      </ul>

    </div>
  </div>
</div>



  
      <div class="fexo-comments comments-post">
    

    

    
    

    

    
    

    

<!-- Gitalk评论插件通用代码 -->
<div id="gitalk-container"></div>

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.css">
<script src="https://cdn.jsdelivr.net/npm/gitalk@1/dist/gitalk.min.js"></script>
<script src="//cdn.bootcss.com/blueimp-md5/2.10.0/js/md5.js"></script>
<script>
const gitalk = new Gitalk({
  clientID: 'e5ff3017bbfe258046d3',
  clientSecret: '9ad425be6fbbe977c47cf51e7f90cced3541a4c3',
  repo: 'fenghuji.github.io',
  owner: 'fenghuji',
  // 在这里设置一下截取前50个字符串, 这是因为 github 对 label 的长度有了要求, 如果超过
  // 50个字符串则会报错.
  // id: location.pathname.split('/').pop().substring(0, 49),
  id: md5(location.pathname),
  admin: ['fenghuji'],
  // facebook-like distraction free mode
  distractionFreeMode: false
})
gitalk.render('gitalk-container')
</script>
<!-- Gitalk代码结束 -->



  </div>

  

  <script type="text/javascript">
  function loadScript(url, callback) {
    var script = document.createElement('script')
    script.type = 'text/javascript';

    if (script.readyState) { //IE
      script.onreadystatechange = function() {
        if (script.readyState == 'loaded' ||
          script.readyState == 'complete') {
          script.onreadystatechange = null;
          callback();
        }
      };
    } else { //Others
      script.onload = function() {
        callback();
      };
    }

    script.src = url;
    document.getElementsByTagName('head')[0].appendChild(script);
  }

  window.onload = function() {
    loadScript('/js/bundle.js?235683', function() {
      // load success
    });
  }
</script>

</body>
</html>
